src/fastfield/serializer/gcd.rs

```diff
@@ -0,0 +1,226 @@
+use fastfield_codecs::{FastFieldDataAccess, FastFieldStats};
+use gcd::Gcd;
+use libdivide::Divider;
+
+pub const GCD_DEFAULT: u64 = 1;
+
+fn compute_gcd(vals: &[u64], base: u64) -> u64 {
+    let mut gcd = (vals[0] - base).gcd(vals[1] - base);
+
+    for el in vals.iter().map(|el| el - base) {
+        gcd = gcd.gcd(el);
+    }
+    gcd
+}
+
+fn is_valid_gcd(vals: impl Iterator<Item = u64>, divider: u64, base: u64) -> bool {
+    if divider <= 1 {
+        return false;
+    }
+    let d = Divider::new(divider).unwrap(); // this is slow
+
+    for val in vals {
+        let val = val - base;
+        if val != (val / &d) * divider {
+            return false;
+        }
+    }
+    true
+}
+
+fn get_samples(fastfield_accessor: &impl FastFieldDataAccess, stats: &FastFieldStats) -> Vec<u64> {
+    // let's sample at 0%, 5%, 10% .. 95%, 100%
+    let num_samples = stats.num_vals.min(20);
+    let step_size = 100.0 / num_samples as f32;
+    let mut sample_values = (0..num_samples)
+        .map(|idx| (idx as f32 * step_size / 100.0 * stats.num_vals as f32) as usize)
+        .map(|pos| fastfield_accessor.get_val(pos as u64))
+        .collect::<Vec<_>>();
+
+    sample_values.push(stats.min_value);
+    sample_values.push(stats.max_value);
+    sample_values
+}
+
+pub(crate) fn find_gcd_from_samples(
+    samples: &[u64],
+    vals: impl Iterator<Item = u64>,
+    base: u64,
+) -> Option<u64> {
+    let estimate_gcd = compute_gcd(samples, base);
```

Comment:
ah right so you do it here... my bad.

I think you can fuse `compute_gcd` and `is_valid_gcd` in one loop. I know I mentionned sampling, but it does not make sense here.